### PR TITLE
feat: use xsnap worker CPU meter and start reporting consumption

### PR DESCRIPTION
**Pull request by michaelfig**
Refs #2322

We implement the same basic policy as in the local worker: refill the meter with a constant value (`1e7` computrons) before every crank and terminate the vat if it consumes more than that in a crank.

There is still a long way to go, but at least now we detect and terminate an xsnap vat upon infinite loops.

### CODE REVIEW

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [inline packages/xsnap/src/xsnap.js]: I'm not happy with this local state to communicate the meter usage to the `vatCancelled` promise.  It's useful functionality, but ugly.

Any suggestions on how to make this better would be appreciated.

File: packages/xsnap/src/xsnap.js
Comment (dckc) [inline packages/xsnap/src/xsnap.js]: for consistency, consider:

```js
const args = [...(snapshot ? ['-r', snapshot] : []),
              ...(meteringLimit ? ['-l', `${meteringLimit}`] : [])]
```

I assume setting `meteringLimit` to 0 is not a useful case.

File: packages/xsnap/src/xsnap.c
Comment (dckc) [inline packages/xsnap/src/xsnap.c]: more globals. Sigh. I guess they're "write once" at least.

File: packages/xsnap/src/xsnap.c
Comment (dckc) [inline packages/xsnap/src/xsnap.c]: again, consistency suggests tabs for indentation

File: packages/xsnap/src/xsnap.js
Comment (dckc) [inline packages/xsnap/src/xsnap.js]: is it important to use `any` rather than `unknown`?

File: packages/xsnap/src/xsnap.js
Comment (dckc) [inline packages/xsnap/src/xsnap.js:124]: `Object` is more or less `any`, yes?

How about:

```
@typedef {{ reply: Uint8Array, meterStats: unknown }} RunResult
```

File: packages/xsnap/src/xsnap.js
Comment (dckc) [inline packages/xsnap/src/xsnap.js]: again, any particular reason for `any` rather than `unknown`?

File: packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js
Comment (dckc) [inline packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js]: The purpose of this comment seems to be to document the type of the return. Any particular reason not to use jsdoc / typescript to do that?

File: packages/xsnap/makefiles/mac/xsnap.mk
Comment (dckc) [inline packages/xsnap/makefiles/mac/xsnap.mk:36]: https://github.com/Moddable-OpenSource/moddable/issues/550 says the metering code is experimental. I wonder whether we should ask @dtribble or somebody before we merge it to master.

**Review by dckc — DISMISSED** [review]
I'm hesitant to move forward with `-DmxMetering=1` while Moddable considers it experimental. I'd like to hear from @dtribble on that, but presuming you're already in communication with him about it, I approve.

I'd really rather not have `any` in the type annotations, but I don't suppose that's critical.

File: packages/xsnap/src/xsnap.js
Comment (dckc) [reply packages/xsnap/src/xsnap.js]: Perhaps @kriskowal would care to weigh in?

There's a lot going on with async and generators and stuff here. I think I had it all swapped in at one time, but I don't now, and this variable is sufficiently local that I'm OK with it.

I'd prefer to see a type annotation.

File: packages/xsnap/makefiles/mac/xsnap.mk
Comment (dckc) [reply packages/xsnap/makefiles/mac/xsnap.mk:36]: @dtribble says ok.

File: packages/xsnap/src/xsnap.js
Comment (warner) [inline packages/xsnap/src/xsnap.js]: Wait, shouldn't the `meterStats` be part of the `OK` that comes back from the worker? Why add a separate message for it? Metering should be measured exactly from the beginning to the end of the crank, and `OK` is how the worker signals the end of the crank, so I'm not in favor of a pathway that suggests that the worker might send metering results at other times.

I think that would also improve the scoping/correlation of the `meterStats` variable, and reduce the changes to the protocol handlers somewhat.

File: packages/xsnap/src/xsnap.c
Comment (warner) [reply packages/xsnap/src/xsnap.c]: I wouldn't say no to a PR that reindented the entire file with spaces...

File: packages/xsnap/src/xsnap.js
Comment (dckc) [reply packages/xsnap/src/xsnap.js]: > Wait, shouldn't the `meterStats` be part of the `OK` that comes back from the worker?

I thought it was, at a higher level... but I only see it in the manager...
https://github.com/Agoric/agoric-sdk/pull/2384/files#diff-73740b92f3f3365578c6c1492419c760b2ec870fa71a269349faca6bfb55f63aR164

I'm not sure how this works without any changes to the supervisor. hmm.

File: packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js
Comment (michaelfig) [reply packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js]: Fixed.

File: packages/xsnap/src/xsnap.c
Comment (michaelfig) [reply packages/xsnap/src/xsnap.c:838]: Done.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js:124]: > `Object` is more or less `any`, yes?

Not when it has any `@property` tags underneath it.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: Now typed.

File: packages/xsnap/src/xsnap.js
Comment (dckc) [inline packages/xsnap/src/xsnap.js]: it's awkward to have the xsnap level know anything about vat-workers.

File: packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js
Comment (michaelfig) [reply packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js]: It was more of a TODO to describe what I wanted for a return result.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: That's more functional, but I have been criticised in the past for using rest arguments with expressions.

A `meteringLimit` of 0 is exactly how one needs to disable metering, so I've updated the conditional accordingly.

File: packages/xsnap/src/xsnap.c
Comment (dckc) [inline packages/xsnap/src/xsnap.c]: how about a test that this matches the package version?

File: packages/xsnap/src/xsnap.c
Comment (michaelfig) [reply packages/xsnap/src/xsnap.c]: > I wouldn't say no to a PR that reindented the entire file with spaces...

I'd suggest creating a `.clang-format` file and enforcing the formatting of our C code.

But that's for the future.  For now, I'm going for consistency.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: I've improved this now.  Will request a re-review when it's ready.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: The worker type is specifically what version of xsnap is in use.

File: packages/xsnap/src/xsnap.c
Comment (michaelfig) [reply packages/xsnap/src/xsnap.c]: Ah, that's where it's supposed to come from?  I'll `-DXSNAP_VERSION=...` in the Makefiles.

File: packages/xsnap/src/xsnap.js
Comment (dckc) [reply packages/xsnap/src/xsnap.js]: ah. different sense of "worker"

File: packages/xsnap/src/xsnap.c
Comment (dckc) [reply packages/xsnap/src/xsnap.c]: > I wouldn't say no to a PR that reindented the entire file with spaces...

Using a style that's inconsistent with Moddable's seems like a bad idea. In particular, it suggests that we're going to be writing / maintaining amounts of code comparable to what they produce, which is not a direction I want to head.

File: packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js
Comment (michaelfig) [reply packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js:6]: The only magic is if a JSDoc type is defined in a file that only has comments, that type is available in all files in the same directory.  `VatPowers` was defined in a different directory, so we need to import them to get them in scope.

File: packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js
Comment (michaelfig) [reply packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js:29]: > `vatManager/types.js` doesn't seem to be reused anywhere else.

It will be soon.  In the next PR, if I'm lucky.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: I followed @warner's recommendation only to propagate these stats when returning OK (`.`).  It did indeed clear up some of the convolutions.

**Review by dckc — COMMENTED** [review]
I think I should withdraw my approval until we have clarity on @warner 's question:
> shouldn't the meterStats be part of the OK that comes back from the worker?

**Review by dckc — COMMENTED** [review]
I was about to approve this, but then ava hangs when I `yarn test` because some xsnap process is hanging around.

**Comment by dckc** [comment]
> ... ava hangs when I `yarn test` because some xsnap process is hanging around.

diagnosis: `makefiles/lin/xsnap.mk` is missing `-DmxMetering=1`

How did this get thru ci?

eek! `packages/xsnap` doesn't seem to get its tests run in ci: https://github.com/Agoric/agoric-sdk/pull/2384/checks?check_run_id=1875722804

**Review by dckc — CHANGES_REQUESTED** [review]
`makefiles/lin/xsnap.mk` is missing `-DmxMetering=1`

and the test-quick ci job is missing `packages/xsnap` altogether!

File: packages/xsnap/src/xsnap.c
Comment (warner) [inline packages/xsnap/src/xsnap.c]: That's a significant compatibility policy we're baking into the snapshot format right there. It means we're claiming 1: we won't make any snapshot-compatibility-breaking changes without also changing the package.json version, and 2: every package.json version change means a compatibility-breaking change. This sort of binding between "marketing" version and compatibility version means we cannot commit a change to the XS code without simultaneously making a release (otherwise we'd be claiming compatibility for the tree before and after the XS commit).

If we're going to move away from a fixed `0.1.0` now, I'd recommend introducing a separate "snapshot compatibility version" (a sequential integer, not a semver string), instead of tying it to the package.json version. (Storing that snapshot compatibility version under a separate key in `package.json` is a fine idea, assuming it doesn't conflict with anything else that parses the file.. I can't think of a better place to put it).

File: packages/xsnap/src/xsnap.c
Comment (warner) [inline packages/xsnap/src/xsnap.c]: as dckc pointed out, this doesn't match the tab/spaces convention of the original, this looks weirdly indented

File: packages/xsnap/src/xsnap.c
Comment (warner) [inline packages/xsnap/src/xsnap.c:918]: I think this is a gratuitous whitespace change

File: packages/xsnap/src/xsnap.js
Comment (warner) [inline packages/xsnap/src/xsnap.js]: `NaN` is a pain to deal with. In this case, it means "unknown", right? As in, the child didn't tell us? Could it be simpler if we require the child to always tell us? And/or use `undefined` to mean "unknown"?

File: packages/xsnap/src/xsnap.js
Comment (warner) [inline packages/xsnap/src/xsnap.js]: I think I understand the need for the semicolon encoding now, but it feels off. Fairly soon, we're going to be returning multiple values from the child, right? Not just CPU usage, but something about peak/current memory allocation? I suspect we'll need something more flexible than `/\.\d+;/`. I guess the deeper issue is the handoff in `xsnap.c` between the JS supervisor code (stashing some "how did it go" value, a string I guess) and the C code that needs to wrap that in a netstring and send it back over the pipe. On one hand, I want the supervisor code to get the metering values, so it can include them next to any status data before it `JSON.stringify`s them into the stashed place. On the other hand, I want the meter to be sampled *after* all JS has finished running, so we aren't looking at meter values while the meter is still in use (which additionally raises the question of what happens if the meter expires after we've sampled it but before we return back into C).

I guess I'm avoiding something: cranks failing for metering reasons is very different than cranks failing for e.g. illegal syscalls. The (JS) supervisor can discover (and report) the latter. But a metering failure will exit the entire JS engine, including the supervisor, so both a metering failure and reporting the meter remaining are things that the C code must do, which means they're necessarily delivered in a different chunk of data than whatever the supervisor provides. Bummer.

Ok so maybe the generalized answer is for each parent-to-child message to result in *two* netstrings in response, glued together. The first can be what the C code reports ("success: no metering violation, here's the remaining meter, or failure because of meter underflow), and the second can be what the JS supervisor code reports ("success: no illegal syscalls"). If the C code sees an error, the second value may be empty (perhaps we synthesize an empty object, just `2:{},`.

**Review by warner — APPROVED** [review]
Tentative approval, please check the comments and see if the recommended changes are ones you're comfortable with. The XSNAP_VERSION thing is the one I'm most concerned about. The message-result protocol is the one I thought the most about but it might not be worth doing anything about.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: > And/or use `undefined` to mean "unknown"?

I will use `null` since that is JSON-serializable.

File: packages/xsnap/src/xsnap.c
Comment (michaelfig) [reply packages/xsnap/src/xsnap.c:918]: It is not.  I'm normalising the spaces in the file to be tabs.

File: packages/xsnap/src/xsnap.c
Comment (michaelfig) [reply packages/xsnap/src/xsnap.c]: I've introduced a `SNAPSHOT_SIGNATURE` manifest constant.

File: packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js
Comment (michaelfig) [reply packages/SwingSet/src/kernel/vatManager/manager-subprocess-xsnap.js:143]: Yes, this PR is a baby step.

File: packages/xsnap/src/netstring.js
Comment (michaelfig) [reply packages/xsnap/src/netstring.js:75]: For a JS generator, the `yield X` results in `{ done: false, value: X }` and `return X` results in `{ done: true, value: Y }`.  The value `Y` is not a signal, it's just an empty value with the same type as `X` so that Typescript doesn't get confused.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: > Ok so maybe the generalized answer is for each parent-to-child message to result in two netstrings in response, glued together.

I had implemented something like that in the first version of this PR (with the `'M'` command code to report metering results), but it was shot down.  Do you want me to resurrect it?

Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: > I suspect we'll need something more flexible than `/\.\d+;/`

There's nothing preventing the evolution of the coupling between `xsnap.c` and `xsnap.js`.  I half-considered `/\.[^\0]*\0/` to encode arbitrary JSON, but opted for the minimal starting point.

Edited: I've chosen instead to specify the meter data as JSON, and to separate it from the other reply with a `\1` character (which is outside of valid JSON characters, so should be more flexible).

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js:18]: Keepers and more flexible metering will need a lot more effort than just passing a meter value into every evaluation.  I don't want to overdesign something we'll have to change again.

File: packages/xsnap/src/xsnap.js
Comment (michaelfig) [reply packages/xsnap/src/xsnap.js]: I've ripped it out in favour of a manifest constant, `METER_TYPE`, which is initially `'xs-meter-1'`.

**Comment by michaelfig** [comment]
> > ... ava hangs when I `yarn test` because some xsnap process is hanging around.
> 
> diagnosis: `makefiles/lin/xsnap.mk` is missing `-DmxMetering=1`

I've added preflight checks to xsnap.c to reject metering unless `-DmxMetering`.

> How did this get thru ci?
> 
> eek! `packages/xsnap` doesn't seem to get its tests run in ci: https://github.com/Agoric/agoric-sdk/pull/2384/checks?check_run_id=1875722804

Added!

**Review by dckc — APPROVED** [review]

**Pull request merged**
